[B] Modified the wording of the second IllegalArgumentException in Player.setScoreboard. Fixes BUKKIT-5149

XHawk87 opened the pull request:
##### The Issue:

The description of the second IllegalArgumentException in Player.setScoreboard (http://jd.bukkit.org/dev/apidocs/org/bukkit/entity/Player.html#setScoreboard(org.bukkit.scoreboard.Scoreboard)) is ambiguous as it could be interpreted as excluding the main (vanilla) scoreboard, as this might not considered to have been "created" by the scoreboard manager.
##### Justification for this PR:

This rewording avoid the confusion of the Main (vanilla) scoreboard not being considered to have been "created" by the scoreboard manager.
##### PR Breakdown:

Changed wording of the javadocs for the second IllegalArgumentException of Player.setScoreboard method from:

IllegalArgumentException - if scoreboard was not created by the scoreboard manager

To:

IllegalArgumentException - if scoreboard was not obtained from the scoreboard manager
##### Testing Results and Materials:

N/A. Minor wording change to javadocs only.
##### Relevant PR(s):

N/A
##### JIRA Ticket:

BUKKIT-5149 - https://bukkit.atlassian.net/browse/BUKKIT-5149

aerouk commented:
Just for future reference, Javadoc PRs don't need as much justification in the commit. Just edit the code and you're done. (I mean you don't have to create a JIRA ticket or follow a PR template for such a small change)

Wolvereness commented:
All `Scoreboards` _are_ created by the scoreboard manager, including the `Scoreboard` that corresponds with the vanilla scoreboard feature.

XHawk87 commented:
Yes, however that is not necessarily clear to someone reading the javadocs as per the reasons outlined above.

Wolvereness commented:
`Scoreboard`s can be obtained from `Player`.`getScoreboard()`, so your change isn't correct.

Your change does not follow javadoc formatting either.

XHawk87 commented:
This ambiguity has caused confusion, otherwise I would not have raised the issue. I also would not have raised the issue if mbaxter hadn't advised me to do so.

Submitting pull requests is recommended as a form of suggestion, I don't necessarily expect them to be pulled. If my suggestion is not acceptable, I can work on another.

I should also point out that if the formatting is incorrect from changing only two words, and staying within the 80 character line limit, then the original formatting is also incorrect, and should be updated.

XHawk87 commented:
@Wolvereness I am fully aware of both the Java Standards for Javadocs as well as the contribution guidelines for Bukkit, and I still do not see any error in the javadoc formatting. Please specify what this error is so that I can correct it for the next PR.

Also, I would appreciate feedback on a more appropriate description of this exception.

Some ideas I had:

if scoreboard was not the main scoreboard nor was it created by the scoreboard manager

if scoreboard was a custom implementation of Scoreboard not created by the scoreboard manager

Or as an additional note:

Note: The Main scoreboard was created by the scoreboard manager so can be used as the given scoreboard

Note: ScoreboardManager.getMainScoreboard() [link] can be used to set the player back to using the vanilla scoreboard feature

SagaciousZed commented:
I am reopening this pending further discussion. As it stands I agree that something should be done.

After reading over it and the surrounding documentation, the exception reason as it currently stands seems to leak implementation that is not explained nor defined elsewhere.

EDIT:
I taked to @Wolvereness and in an attempt clarify his stance on the current change. He does not think the proposed change is correct, because it is to narrow. Scoreboards may be obtained from other classes that are created by the Scoreboard.

XHawk87 committed:
Added clarification to Player.setScoreboard to ensure users know how to
return a player to using the main scoreboard from a custom scoreboard and
alleviate any confusion caused by the second IllegalArgumentException
description

XHawk87 commented:
Thanks @SagaciousZed. 

I have updated the description to instead include a paragraph explaining how to return to using the main scoreboard after a custom one had been set. I believe this should clear up several unanswered questions that a user may have about this method, as well as reassure them that the main scoreboard is not a special case that would be excluded by the second IllegalArgumentException.

Any constructive feedback on this new description would be appreciated, to ensure that it is factually correct and complete, as well as unambiguous in meaning, and in line with Bukkit guidelines for JavaDocs.

Wolvereness commented:
This improperly identifies the goal of the scoreboard API. Scoreboards are a stat tracking mechanism, not a general UI (yes, I know how plugins use it, and I may be guilty myself). This appears to be documenting it as if its a UI system (which is wrong), but we don't support that type of behavior.

References like _remove_ are better described as _revert_. There is no concept of remove, nor should references to it be added; it makes no sense. 

The terminology is action oriented (_set_), better represented as state driven (_has_). Players don't need things to be set when they log into a server; it just has it. Changing the scoreboard is a state transition, not a display system.

XHawk87 commented:
How about this?

```
A player always has an active scoreboard, even if it has no objectives. It is
possible to switch from one custom scoreboard directly to another,
however, if the goal is to revert to using the default scoreboard feature,
the {@link org.bukkit.scoreboard.ScoreboardManager#getMainScoreboard() main
scoreboard} may be used.
```

Wolvereness commented:
It's too long of a sentence, but I like the sentiments.

XHawk87 commented:
Broken up, and added more detail.

```
A player always has an active scoreboard, even if it has no objectives. 
Initially, this is the global scoreboard for the server which can be 
manipulated through the /scoreboard command. The active scoreboard can be 
switched to any custom scoreboard created through the {@link 
org.bukkit.scoreboard.ScoreboardManager}. This can also be used to revert to 
using the global scoreboard from the {@link 
org.bukkit.scoreboard.ScoreboardManager#getMainScoreboard()} method.
```

Wolvereness commented:
Now, I don't like where you went with that. If it's worth describing how to create scoreboards, it's worth noting that scoreboards don't have to come 'through' the `ScoreboardManager`, rather, the initial creation must be from the `ScoreboardManager`.

XHawk87 commented:
Perhaps this then?

```
A player always has an active scoreboard, even if it has no objectives. 
Initially, this is the {@link 
org.bukkit.scoreboard.ScoreboardManager#getMainScoreboard() main scoreboard} 
for the server which can be manipulated through the /scoreboard command. The 
active scoreboard can be switched to any scoreboard initially created by the 
{@link org.bukkit.scoreboard.ScoreboardManager}, including the main 
scoreboard.
```
